posthog/temporal/batch_exports/temporary_file.py

```diff
@@ -0,0 +1,523 @@
+                quotechar=self.quote_char,
+                escapechar=self.escape_char,
+                quoting=self.quoting,
+                lineterminator=self.line_terminator,
+            )
+
+        return self._csv_writer
+
+    def _write_record_batch(self, record_batch: pa.RecordBatch) -> None:
+        """Write records to a temporary file as CSV."""
+        self.csv_writer.writerows(record_batch.to_pylist())
+
+
+class ParquetBatchExportWriter(BatchExportWriter):
+    """A `BatchExportWriter` for Apache Parquet format.
+
+    We utilize and wrap a `pyarrow.parquet.ParquetWriter` to do the actual writing. We default to their
+    defaults for most parameters; however this class could be extended with more attributes to pass along
+    to `pyarrow.parquet.ParquetWriter`.
+
+    See the pyarrow docs for more details on what parameters can the writer be configured with:
+    https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetWriter.html
+
+    In contrast to other writers, instead of us handling compression we let `pyarrow.parquet.ParquetWriter`
+    handle it, so `BatchExportTemporaryFile` is always initialized with `compression=None`.
+
+    Attributes:
+        schema: The schema used by the Parquet file. Should match the schema of written RecordBatches.
+        compression: Compression codec passed to underlying `pyarrow.parquet.ParquetWriter`.
+    """
+
+    def __init__(
+        self,
+        max_bytes: int,
+        flush_callable: FlushCallable,
+        schema: pa.Schema,
+        compression: str | None = "snappy",
+    ):
+        super().__init__(
+            max_bytes=max_bytes,
+            flush_callable=flush_callable,
+            file_kwargs={"compression": None},  # ParquetWriter handles compression
+        )
+        self.schema = schema
+        self.compression = compression
+
+        self._parquet_writer: pq.ParquetWriter | None = None
+
+    @property
+    def parquet_writer(self) -> pq.ParquetWriter:
+        if self._parquet_writer is None:
+            self._parquet_writer = pq.ParquetWriter(
+                self.batch_export_file,
+                schema=self.schema,
+                # Compression *can* be `None`.
+                compression=self.compression,
+            )
+        return self._parquet_writer
+
+    def ensure_parquet_writer_is_closed(self) -> None:
+        """Ensure ParquetWriter is closed as Parquet footer bytes are written on closing."""
+        if self._parquet_writer is None:
+            return
+
+        self._parquet_writer.writer.close()
+        self._parquet_writer = None
+
+    @contextlib.asynccontextmanager
+    async def open_temporary_file(self):
+        """Ensure underlying Parquet writer is closed before flushing and closing temporary file."""
+        async with super().open_temporary_file():
+            try:
+                yield
+            finally:
+                self.ensure_parquet_writer_is_closed()
```

Comment:
Mostly wanted to give this a name. I agree I don't think this will be finalized anywhere else, so happy to move it back in here until the need arises to have a method (and if it arises).